chia/wallet/wallet_state_manager.py

```diff
@@ -651,10 +663,94 @@ async def determine_coin_type(
+                    clawback_coins[coin.coin] = metadata
+                    if len(clawback_coins) >= self.config.get("auto_claim", {}).get("btach_size", 50):
+                        await self.claim_clawback_coins(clawback_coins, tx_fee)
+                        clawback_coins = {}
+        if len(clawback_coins) > 0:
+            await self.claim_clawback_coins(clawback_coins, tx_fee)
+
+    async def claim_clawback_coins(self, clawback_coins: Dict[Coin, ClawbackMetadata], fee: uint64) -> List[bytes32]:
+        assert len(clawback_coins) > 0
+        coin_spends: List[CoinSpend] = []
+        message_list: List[bytes32] = []
+        for coin in clawback_coins.keys():
+            message_list.append(coin.name())
+        message: bytes32 = std_hash(b"".join(message_list))
+        spent_coins = []
+        for coin, metadata in clawback_coins.items():
+            recipient_puzhash = metadata.recipient_puzzle_hash
+            sender_puzhash = metadata.sender_puzzle_hash
+            if metadata.is_recipient:
+                derivation_record = await self.puzzle_store.get_derivation_record_for_puzzle_hash(recipient_puzhash)
+            else:
+                derivation_record = await self.puzzle_store.get_derivation_record_for_puzzle_hash(sender_puzhash)
+            if derivation_record is None:
+                self.log.warning(f"You are not able to spend coin {coin.name().hex()}")
+                continue
+            inner_puzzle: Program = self.main_wallet.puzzle_for_pk(derivation_record.pubkey)
+            inner_solution: Program = self.main_wallet.make_solution(
+                primaries=[
+                    {
+                        "puzzlehash": derivation_record.puzzle_hash,
+                        "amount": uint64(coin.amount),
+                        "memos": [],
+                    }
+                ],
+                coin_announcements=None if len(coin_spends) > 0 or fee == 0 else {message},
+            )
+            coin_spends.append(generate_clawback_spend_bundle(coin, metadata, inner_puzzle, inner_solution))
+            spent_coins.append(coin.name())
+        spend_bundle = await self.main_wallet.sign_transaction(coin_spends)
+        if fee > 0:
+            chia_tx = await self.main_wallet.create_tandem_xch_tx(
+                fee, Announcement(coin_spends[0].coin.name(), message)
+            )
+            assert chia_tx.spend_bundle is not None
+            spend_bundle = SpendBundle.aggregate([spend_bundle, chia_tx.spend_bundle])
+            chia_tx = dataclasses.replace(chia_tx, spend_bundle=None)
+            await self.add_pending_transaction(chia_tx)
+        # We don't want to mess up tx record here so just push the bundle
+        await self.wallet_node.push_tx(spend_bundle)
+        for coin_spend in coin_spends:
+            # This will make sure we don't double spend
+            await self.coin_store.set_spent(coin_spend.coin.name(), uint32(0))
```

Comment:
I expect we can make this quite a lot faster by having a batch-set-spent function on the coin store (I sped up the full node coin store that way).

However, it looks a bit suspicious to mark these coins as spent just because we've submitted a transaction to the mempool. The transaction might be rejected or evicted from the mempool. I would have thought we would mark them as pending, or something like that.